### PR TITLE
Sockets - Jansen

### PR DESCRIPTION
**Pull request by JansenMartin**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method in Work called `top_media`.  It returns a collection of Work objects that are a) the same category, b) have a vote count more than 0, and c) are within the top ten vote count among works.
Describe how you approached testing that model method. What edge cases did you come up with? | I wrote tests to make sure a) it returned Work objects, b) that all Works inside the collection have more than 0 votes, and c) it returned an empty collection if there were no Works in the database.  The last case is an edge case.
What are `session` and `flash`? What is the difference between them? |  The data stored in `session` persists for an infinite number of request-response cycles because it's stored inside the client's cookies.  Meanwhile, the data inside `flash` persists for only 1-2 response-request cycles.  The latter is ideal for displaying messages based on user actions and status codes.  The former is ideal for keeping track of logged-in users.   
What was one thing that you gained more clarity on through this assignment? | I became much more comfortable switching between Git branches, as well as running tests before and after merging branches.  
What is the Heroku URL of your deployed application? | [Here](https://jansen-media-ranker.herokuapp.com/)
Do you have any recommendations on how we could improve this project for the next cohort? |  It might be cool to have more resources about how to generate fixture data based off of the seeded database.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      @user = User.new(user_params)
+      unless @user.save
+        flash[:status] = :error
+        flash[:message] = "Please try again."
+        render :login_form, status: :bad_request
+        return
+      else
+        session[:user_id] = User.find_by(id: @user.id)
+        flash[:status] = :success
+        flash[:message] = "Nice to meet you, #{username}!"
```
Comment (droberts-sea) [inline app/controllers/users_controller.rb:36]: On line 34, you attempt to store the entire `User` object in the session, instead of its ID. This line would work better as

```ruby
session[:user_id] = @user.id
```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    @user = User.find_by(username: username)
+
+    if @user
+      session[:user_id] = @user.id
+      flash[:status] = :success
+      flash[:message] = "Welcome back, #{username}!"
+    else
+      @user = User.new(user_params)
+      unless @user.save
+        flash[:status] = :error
```
Comment (droberts-sea) [inline app/controllers/users_controller.rb:29]: Minor style note: typically if we're going to need both the true and false branches of a conditional, we use an `if` instead of an `unless`.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  def show
+    unless @work
+      head :not_found
+    end
+  end
+
+  def edit
+    unless @work
+      head :not_found
+    end
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:39]: This work is repeated across several methods here. You already have a controller filter to find the work, could you include this error checking in that?

File: app/models/vote.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work, counter_cache: true
+
+  validates :work_id, uniqueness: { scope: :user_id,
+                                    message: "User cannot vote for the same work twice." }
```
Comment (droberts-sea) [inline app/models/vote.rb:6]: Good work getting this tricky `uniqueness` scope and the `counter_cache` figured out!

File: config/routes.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  root "homepages#index"
+
+  resources :works do
+    resources :votes, only: [:create]
+  end
+
+  resources :users, only: [:index, :show]
```
Comment (droberts-sea) [inline config/routes.rb:9]: Good work generating only the routes you need here.

File: app/views/homepages/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    <h2>Media Spotlight: <%= @media_spotlight.title%> by <%= @media_spotlight.creator%> </h2>
+  <% else %>
+    <h2>No Media Spotlight</h2>
+  <% end %>
+  
+  <%# MOVIES %>
+  <h2>Top Movies</h2>
+  <section>
+   <% if @top_movies.count > 0 %>
+      <ul>
```
Comment (droberts-sea) [inline app/views/homepages/index.html.erb:19]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: test/controllers/homepages_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    get root_path
+
+    # Assert
+    must_respond_with :ok
+  end
+
+  it "renders even when there aren't any works" do
+    # Arrange
+    # Destroy everything in the database
+    Work.destroy_all
```
Comment (droberts-sea) [inline test/controllers/homepages_controller_test.rb:20]: I like this test case - with the spotlight this could very plausibly break!

File: test/controllers/users_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+      # Assert
+      must_respond_with :ok
+    end
+  end
+
+  describe "login" do
+    it "adds a new user to the database" do
+      # Arrange
+      user_data = {
```
Comment (droberts-sea) [inline test/controllers/users_controller_test.rb:21]: Good work getting the all 3 interesting test cases here.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene |yes
Comprehension questions |yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Views are well-organized (DRY, use of semantic HTML, use of partials) | mostly - see inline
Errors are reported to the user | yes
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | mostly - missing tests for the `Work#spotlight` method
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | missing tests for voting, otherwise this looks good
**Wave 1 - Media** |
Splash page shows the three media categories | yes (only if they have votes)
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | mostly - logging in a new user doesn't work (see inline)
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** | Great job overall! Your implementation matches the demo site very closely, and I would say the learning goals for this assignment were definitely met. There are a few places where things could be cleaned up, which I've tried to outline below, but in general I am quite happy with this submission. Keep up the hard work!
